File: extension.ts
/**
 * CodeMore
 */
//% color=#0082FB weight=96 icon="\uf294"
namespace CodeMore {

    /**
    * Starts BLE services for Scratch Microbit-More extension.
    */
    //% blockId=CodeMore_startCodeMoreService block="start Microbit-More service"
    //% shim=CodeMore::startService
    export function startService():void {
        console.log("Microbit-More started");
    }

    let initialized = false;
    let onDisplayTextCommandBody: (cmd: string) => void;

    function init() {
        if (initialized) return;
        initialized = true;
        onCodemoreEvent(handleCodemoreEvent);
    }

    /**
     * Used internally by the library.
     */
    //% block
    //% deprecated=true blockHidden=1
    //% shim=CodeMore::onCodemoreEvent
    function onCodemoreEvent(body: () => void): void {
        console.log("shim=custom::onCodemoreEventImpl")
    }

    function handleCodemoreEvent(): void {
        if (onDisplayTextCommandBody) {
            let cmd: string = getLastDisplayTextCommand();
            onDisplayTextCommandBody(cmd);
        }
    }

    /**
     * on display text command received
     */
    //% blockId=CodeMore_onDisplayTextCommand block="on display text command received" blockGap=16
    //% draggableParameters=reporter
    export function onDisplayTextCommand(body: (cmd: string) => void) {
        init();
        onDisplayTextCommandBody = body;
    }

}

File: extension.cpp
#include "pxt.h"

#if MICROBIT_CODAL
#include "MbitMoreServiceV2.h"
#else
#include "MbitMoreService.h"
#endif

#define UPDATE_PERIOD 11
#define NOTIFY_PERIOD 101

/**
 * CodeMore
 */
//% color=#0082FB weight=96 icon="\uf294"
namespace CodeMore {
    
    MbitMoreService* _pService = NULL;
    MicroBitEvent _evt(DEVICE_ID_CODEMORE, DEVICE_CODEMORE_EVT_DIPSPLAY_TEXT_CMD);

    ManagedString _lastDisplayTextCommand("");

    /**
     * Get last display text command
     */
    //% block
    //% deprecated=true blockHidden=1 
    String getLastDisplayTextCommand() {
        return PSTR(_lastDisplayTextCommand);
    }

    int	displayTextCommand(MicroBit &e, ManagedString &mstr) {
        _lastDisplayTextCommand = mstr;
        _evt.fire();
        return 0; // handled!
    }

    void update() {
        while (NULL != _pService) {
            _pService->update();
            fiber_sleep(UPDATE_PERIOD);
        }
    }

    void notifyScratch() {
        while (NULL != _pService) {
            // notyfy data to Scratch
            _pService->notify();
            fiber_sleep(NOTIFY_PERIOD);
        }
    }

    //%
    void startService() {
        if (NULL != _pService) return;

        _pService = new MbitMoreService(uBit);
        _pService->assigneCallbackDisplayTextCommand(displayTextCommand);
        create_fiber(update);
        create_fiber(notifyScratch);
    }
    
    //%
    void onCodemoreEvent(Action body) {
        registerWithDal(DEVICE_ID_CODEMORE, DEVICE_CODEMORE_EVT_DIPSPLAY_TEXT_CMD, body);
    }

}
